### ISSUE TITLE
Group-2: Refactor Clean End-to-End Modular Pipeline with Nested Cross-Validation

### ISSUE THREAD

**Issue by project-delphi** [ccm-bioinfo/Camda24_resistance#11]
**Objective**:
Develop a modular and reusable end-to-end pipeline that handles data preprocessing, model training, cross-validation, and evaluation on test data. Use nested cross-validation to ensure unbiased performance estimates.

#### Tasks to Complete

1. **Organize Code into Modular Components**
   - [ ] **Data Preprocessing Module**:
     - Create a separate module for data preprocessing (e.g., handling missing data, scaling, encoding).
     - Ensure the preprocessing pipeline is flexible and can be integrated with different datasets.
   - [ ] **Model Definition Module**:
     - Implement a module that defines multiple models to be used in the pipeline (e.g., `RandomForest`, `SVM`, `XGBoost`).
     - Add a function to select models and hyperparameters for tuning.
   - [ ] **Evaluation and Metrics Module**:
     - Create a module for performance evaluation metrics (e.g., accuracy, F1-score, RMSE).
     - Allow easy switching between different scoring functions, especially for regression vs. classification tasks.

2. **Construct a Modular Pipeline Using Sklearn**
   - [ ] Use `Pipeline` from `sklearn.pipeline` to combine preprocessing steps and model training.
   - [ ] Ensure the pipeline can handle cross-validation and hyperparameter tuning.
   - [ ] Modularize the code so that each component (preprocessing, modeling, evaluation) is reusable.

3. **Nested Cross-Validation**
   - [ ] Implement nested cross-validation using `GridSearchCV` or `RandomizedSearchCV` for hyperparameter tuning inside the cross-validation loop.
   - [ ] Set up outer cross-validation to assess model generalization, while the inner cross-validation handles hyperparameter tuning.
   - [ ] Ensure proper handling of training and validation sets within the nested cross-validation.

4. **Run the Pipeline on Training Data**
   - [ ] Train the pipeline on the training data using cross-validation to tune hyperparameters.
   - [ ] Capture performance metrics from the nested cross-validation (e.g., mean score, standard deviation).

5. **Evaluate on Test Data**
   - [ ] After cross-validation, evaluate the best model on the test data.
   - [ ] Report the final performance on the test dataset, including metrics like accuracy, precision, recall, or RMSE depending on the task.

6. **Performance Reporting**
   - [ ] Create a report summarizing cross-validation results:
     - Mean and standard deviation of cross-validation scores.
     - Final performance on the test dataset.
     - Feature importance or any model insights.
   - [ ] Include visualizations like confusion matrices, ROC curves, or feature importance plots.

#### Example Code Structure

```python
from sklearn.pipeline import Pipeline
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import GridSearchCV, cross_val_score, train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import accuracy_score

# Define preprocessing steps
preprocessor = Pipeline(steps=[
    ('scaler', StandardScaler())
])

# Define model
model = RandomForestClassifier()

# Create full pipeline
pipeline = Pipeline(steps=[
    ('preprocessor', preprocessor),
    ('model', model)
])

# Set up grid search for hyperparameter tuning
param_grid = {
    'model__n_estimators': [50, 100],
    'model__max_depth': [None, 10, 20]
}
grid_search = GridSearchCV(pipeline, param_grid, cv=5)

# Split the data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

# Nested cross-validation
cross_val_scores = cross_val_score(grid_search, X_train, y_train, cv=5)

# Train and evaluate on test data
grid_search.fit(X_train, y_train)
test_predictions = grid_search.predict(X_test)
test_accuracy = accuracy_score(y_test, test_predictions)

# Report results
print(f'Cross-validation scores: {cross_val_scores}')
print(f'Final test accuracy: {test_accuracy}')

**Comment by project-delphi** [comment]
reset to clean modules with tested code.
